### PR TITLE
openvpn-policy-routing: service to route traffic via specific gateways (WAN or multiple OpenVPN tunnels).

### PR DESCRIPTION
**Pull request by stangri**
Signed-off-by: Stan Grishin <stangri@melmac.net>

Maintainer: me / @stangri
Compile tested: (ipq806x, Linksys EA8500, LEDE 17.01 snapshot)
Run tested: (ipq806x, Linksys EA8500, LEDE 17.01 snapshot)

Description:
This service supersedes vpnbypass by providing ability to target not only WAN gateway, but specific OpenVPN tunnels gateways with both IP/Port-based policies and domain-based policies (requires dnsmasq-full for latter).
Along with companion package (luci-app-openvpn-policy-routing) provides an easy to use way to route traffic when 1 or more OpenVPN tunnels are used.

Please provide feedback on the code.

### CODE REVIEW

**Comment by hnyman** [comment]
Minor comment, but I don't like the "sed magic" for the init script where you place the package name and version to the init script in the prepare step. Looks rather unnecessary especially regarding the name. (Printing the package version to the logs can be helpful, though)

This will replace the older vpnbypass package? If that is the case, then I think that the vpnbypass it should be removed rather soon, before it picks a larger user base.

**Comment by hnyman** [comment]
That "minor comment" is apparently a major one. You are using sed to patch the source files directly in the main source directory, instead of doing that stuff in a separate package build directory.

That seems to lead into problems, as the next "git pull" source update for the feeds fails.
You have already used that in vpnbypass, and now phase2 buildbot is failing with 
```
Updating feed 'packages' from 'https://git.lede-project.org/feed/packages.git' ...
From https://git.lede-project.org/feed/packages
   cc5f005..d156404  master     -> origin/master
error: Your local changes to the following files would be overwritten by merge:
    net/vpnbypass/files/vpnbypass.init
Please, commit your changes or stash them before you can merge.
Aborting
```
That part needs to be removed.

**Comment by stangri** [comment]
Hello Hannu, thank you for the feedback and sorry about the vpnbypass. What (besides running sed on a file in the tree) should I have done differently there? There were a few changes and I'm not sure if I described each of them in the message -- was that contributing factor?

I'd like to keep vpnbypass around as for majority of users I imagine it's an easier tool to use.

Regarding the sed, I get that I've modified the code in the tree and that broke things, but coming back to your earlier statement about sed:
1. I understand now that the files should have been modified in SRC folder outside of the tree, I will look into that. Sadly most of the information I could google about the makefile doesn't seem to apply to OpenWrt/LEDE.
2. Is sed the bad tool for this? Are there make-provided means of stamping the version in the source code?
3. I'll try to submit the fixed makefile with the correct way of stamping version for vpnbypass first and will then apply the same to openvpn-policy-routing.

**Comment by hnyman** [comment]
Patching the version to the init file at the compilation time just adds complexity. If you want the version info there, just edit the file manually at the same time as you edit the version in the Makefile. (That is the way e.g. ddns-scripts maintains its version info.)

You have only scripts, so adding a source copying step and prepare step should be unnecessary...

But if you insist in editing one script file at build time, you need to do something like in https://github.com/openwrt/packages/blob/master/net/cgi-io/Makefile so that the files are copied to PKG_BUILD_DIR and then modified there, and finally copied from there at the install step.

Editing a source file with sed is ok in principle, but doing it to the original package source at the build time will invalidate that source file from git perspective. (So, the next git pull will fail,  also in all private builders' local repos who have included vpnbypass in their builds so far.)

Openwrt/LEDE package Makefile is not a simple usual Makefile, but includes the local abstraction layer that defines the package handling etc.. So practically the only knowledge source is Openwrt & LEDE wikis plus looking at the Makefiles in the repos. https://wiki.openwrt.org/doc/devel/packages

No hurry, as buildbots are still choking from the effects of the previous commit...

(Jow has prepared in his staging repo a cleanup script for buildbots to prevent damage from rogue packages in future and to cleanup the current damage, but that is not applied yet, so the phase2 buildbot status is pretty red :-( 

Ps. And you need to delete that screenshot PNG file from the git repo. It makes no sense that everybody downloads it. If you want that to be visible in the README, include it in your local Github fork and use that as the link in README. But do not include the actual screenshot in the official source repo.

**Comment by stangri** [comment]
Well, at least I can be useful as a bad example. ;)

Regarding the version numbering -- thank you for providing a reference, I'll submit the modified code for your review.

Regarding the screenshot -- I'm concerned that if it was hosted externally it might not be served reliably as things can change outside of the official source repo, but if the screenshot is there it will be safely served. Please advise.

**Comment by danrl** [comment]
I understand that you want to provide screenshots. It's just that it cloggs up builder's disk space for nothing. I have multiple builds on one machine and every instance has a copy of all involved source repos. Every unnecessary byte adds up to diskspace consumption, not cheap on SSD.

What I came up when I had the idea to provide screenshots and educational material, I wrote blog posts on my personal blog and linked them. Also, writing the info and the screenshots to the lede wiki helps spreading the word/pic. I fully understand your desire to have the screenshot attached. But please do that outside the source repository if possible.

**Comment by stangri** [comment]
@hnyman, @danrl -- got it guys, I've already moved the screenshots into my own screenshot repo.

File: net/openvpn-policy-routing/files/openvpn-policy-routing.init [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+	iptables_reset; { modprobe xt_set; modprobe ip_set; modprobe ip_set_hash_ip; } >/dev/null 2>&1;
+	config_load network; config_foreach process_interfaces interface create;
+	config_load openvpn-policy-routing; config_foreach process_policy policy; config_foreach process_policy policy-ipv6 "ipv6";
+	if [ uci -q get openvpn-policy-routing.@domain-policy[] >/dev/null 2>&1 ]; then
+		output "routing domain-based policies" 2;
+		config_foreach process_domain_policy domain-policy
+		[ -z "$d_status" ] && output " $__ok__\n" 2 || output " $__fail__\n$__error__:$d_status\n" 2
+	elif f_dnsmasq "ipset"; then
+		output "routing domain-based policies with dnsmasq" 2;
+		if [ ! -e /var/lock/opkg.lock ] && ! opkg list_installed | grep 'dnsmasq-full' >/dev/null 2>&1 ; then
```
Comment (stangri) [inline net/openvpn-policy-routing/files/openvpn-policy-routing.init]: Is there a more elegant (and ideally not opkg-dependant) way to determine that dnsmasq-full is installed and/or that the installed dnsmasq includes ipset support?

Comment (hnyman) [inline net/openvpn-policy-routing/files/openvpn-policy-routing.init]: Using "dnsmasq --version". It lists all optional compile-time features.

```
root@OpenWrt2:~# dnsmasq --version
Dnsmasq version 2.76  Copyright (c) 2000-2016 Simon Kelley
Compile time options: IPv6 GNU-getopt no-DBus no-i18n no-IDN DHCP no-DHCPv6 no-Lua TFTP no-conntrack no-ipset no-auth no-DNSSEC no-ID loop-detect inotify

```

Comment (stangri) [inline net/openvpn-policy-routing/files/openvpn-policy-routing.init]: Thank you! I was sure I saw it somewhere, but I've tried dnsmasq --help and didn't gather I should have tried -v from it.

**Comment by stangri** [comment]
I'm getting closer to finalizing this package. Latest code should support IPv6 and I would very much welcome review/feedback from people actively using IPv6, while I'm struggling to get going with it.

**Comment by hnyman** [comment]
> I'm getting closer to finalizing this package.

Has it stabilised enough for updating the PRs, and merging it in?

**Comment by stangri** [comment]
No, not yet. :( Now that my OpenVPN provider supports IPv6 I've been meaning to set up 6in4 on my router to test the IPv6 functionality, but haven't had time to do that.
So as it is, I don't have any success reports on IPv6 testing of this package.
If the existence of this PR is annoying, it can be closed and I will resubmit after IPv6 testing.

**Comment by hnyman** [comment]
No need to cloe it, you can leave it open, but you might update it (and Luci) with your current versions. Not everybody finds your personal repo.

**Pull request closed without merging**